node/src/client/mod.rs

```diff
@@ -181,6 +201,72 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         }));
     }
 
+    /// Initializes transaction verification.
+    fn initialize_transaction_verification(&self) {
+        // Start the transaction verification loop.
+        let node = self.clone();
+        self.handles.lock().push(tokio::spawn(async move {
+            loop {
+                // If the Ctrl-C handler registered the signal, stop the node.
+                if node.shutdown.load(Relaxed) {
+                    info!("Shutting down transaction verification");
+                    break;
+                }
+
+                // Sleep briefly to allow transactions to be validated.
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+
+                // Determine the length of the queue.
+                let queue_len = node.transaction_queue.lock().len();
+
+                // If we don't have any transactions to verify, try again later.
+                if queue_len == 0 {
+                    continue;
+                }
+
+                // Determine how many transactions we want to verify.
+                let num_transactions = VERIFICATION_CONCURRENCY_LIMIT.min(queue_len);
+
+                // Check if we have room to verify the transactions and update the counter accordingly.
+                let previous_verification_counter = node.verification_counter.fetch_update(Relaxed, Relaxed, |c| {
+                    // If we are already verifying sufficient transactions, don't verify any more for now.
+                    if c >= VERIFICATION_CONCURRENCY_LIMIT {
+                        None
+                    // If we have space to verify more transactions, verify as many as we can.
+                    } else {
+                        // Consider verifying the full transaction queue, but limit to the concurrency limit.
+                        Some((c + queue_len).min(VERIFICATION_CONCURRENCY_LIMIT))
+                    }
+                });
+
+                // Determine how many transactions we cán verify.
+                let num_transactions = match previous_verification_counter {
+                    // Determine how many transactions we can verify.
+                    Ok(previous_value) => num_transactions.saturating_sub(previous_value),
+                    // If we are already verifying sufficient transactions, don't verify any more for now.
+                    Err(_) => continue,
+                };
+
+                // For each transaction, spawn a task to verify it.
+                for _ in 0..num_transactions {
+                    if let Some((_, (peer_ip, serialized, transaction))) = node.transaction_queue.lock().pop_lru() {
```

Comment:
I think the cache should be unlocked just once for the entirety of the loop, i.e. we should bind `node.transaction_queue.lock()` before it

Reply:
Sure: 321cd023e
Just so our mental models are aligned: given that we want it to take the "oldest seen" transactions, this won't change the order of transactions which are validated right? But just be more robust and perhaps speedy.

Added this to ensure we will really process by oldest tx first: 0d383e525